stately/src/otherNativeMain/kotlin/co/touchlab/stately/concurrency/Lock.kt

```diff
@@ -14,15 +26,33 @@ actual class Lock actual constructor() {
   }
 
   actual fun tryLock():Boolean {
-    return lock.compareAndSet(0, 1)
+    return when (lock.compareAndSwap(0, Thread.id)) {
+      Thread.id -> {
+        // We already have the lock, we are just re-entering.
+        reenterCount.increment()
+        true
+      }
+      0 -> {
+        // Another thread didn't have the lock so we just took it.
+        assert(reenterCount.value == 0)
+        true
+      }
+      else -> false
+    }
   }
 
-  private fun spinLock(){
-    while (!lock.compareAndSet(0, 1)){}
+  private fun spinLock() {
+    while (!tryLock()) {}
   }
 
   private fun spinUnlock(){
-    if(!lock.compareAndSet(1, 0))
-      throw IllegalStateException("Lock must always be 1 in unlock")
+    // Because this is re-entrant we should only unlock if the count is 0.
+    if (reenterCount.value > 0) {
```

Comment:
this should be some kind of `compareAndSwap`

Reply:
Why do you say that?

Reply:
Actually I'm not sure this is the right suggestion, but the point is that 2 threads could get here and both see this when the count 1 one, go through this condition, and then decrement the value below 0

Reply:
Yeah, this is a somewhat naive implementation because it expects this only to be called by a thread that has already gotten a lock.

Reply:
If it were called by someone that didn't have a lock, it would break things.

Reply:
I could throw an exception if it is called by a thread other than the one holding the lock.

Reply:
how about 
```kotlin
val reenterCount = reenterCount.decrementAndGet()
if (reenterCount > 0) {
    return;
}
assert(reenterCount == 0, "Lock unlocked too many times")
lock.set(0)
```

Reply:
The real problem is just that this method could be called by another thread. There is the much larger problem in that case that the lock would be broken.

I think just adding this to the top of the method solves both concerns:
`assert(lock.value == Thread.id)`

Reply:
I think this requires my suggested modification to lock

Reply:
The lock should be usable from other threads though. This is valid imo (scrappy implementation of a future):
```kotlin
val lock = Lock()
lock.lock()
performAsync {
    // another thread
    lock.unlock()
}
lock.lock()
lock.unlock()
```

Reply:
I don't know... That sounds dicey to me. Most of my reading about locks, particularly recursive ones, the lock is expected to be unlocked by the same thread that locked it to begin with.
This would probably be bad from the point of consistency as well. NSLock itself says that it will cause undefined behavior if you attempt to unlock from a different thread.

Reply:
Ugh that's frustrating. Okay. pthread_unlock doesn't explicitly forbid this. It'd be great if this library could be built with better locks. I'd like to propose a minor refactor then. Rename the expect/actual class to `LockImpl`, and then `Lock` will wrap LockImpl. LockImpl will be required to provide a method called `isCurrentThreadOwner()` which will return true, if the current thread owns the lock. Then, we can implement unlock to check this and throw. If the behavior is undefined on iOS, we should make it difficult to write common code that runs into this.

Reply:
I don't disagree with the idea, but I don't want to do that as part of this PR. Seems out of scope since this is only meant to make the native lock re-entrant. Once this gets merged, I am happy to work on further improvements.

Reply:
Sounds good!

Reply:
So to summarize, move forward with not allowing unlock from another thread.

Reply:
Can we assert earlier in the `unlock` code that we're on the correct thread?

Reply:
You want me to move the assert into `unlock` rather than `spinUnlock`?

Reply:
Whoops typo. I just mean earlier in the `spinUnlock` code

Reply:
The assert is in the first line of the method...

Reply:
Oh lol. Sorry conflated my sample implementation with what you wrote.